pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,410 @@
+
+    >>> with qml.tape.QuantumTape() as tape:
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     qml.expval(qml.PauliZ(0))
+    >>> gradient_tapes, fn = qml.gradients.hadamard_grad(tape)
+    >>> gradient_tapes
+    [<QuantumTape: wires=[0, 1], params=3>,
+     <QuantumTape: wires=[0, 1], params=3>,
+     <QuantumTape: wires=[0, 1], params=3>]
+
+    This can be useful if the underlying circuits representing the gradient
+    computation need to be analyzed.
+
+    The output tapes can then be evaluated and post-processed to retrieve
+    the gradient:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> fn(qml.execute(gradient_tapes, dev, None))
+    (array(-0.3875172), array(-0.18884787), array(-0.38355704))
+
+    .. note::
+
+        ``hadamard_grad`` will decompose the operations that are not in the list of supported operation.
+
+        - ``pennylane.RX``
+        - ``pennylane.RY``
+        - ``pennylane.RZ``
+        - ``pennylane.Rot``
+        - ``pennylane.PhaseShift``
+        - ``pennylane.U1``
+        - ``pennylane.CRX``
+        - ``pennylane.CRY``
+        - ``pennylane.CRZ``
+        - ``pennylane.IsingXX``
+        - ``pennylane.IsingYY``
+        - ``pennylane.IsingZZ``
+
+        The expansion will fail if a suitable decomposition in terms of supported operation is not found,
+        an error is then raised. The number of trainable parameters can potentially increase due to the
+        decomposition.
+
+    """
+    if any(isinstance(m, VarianceMP) for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of variance with hadamard test gradient is not implemented yet."
+        )
+    if any(isinstance(m, (StateMP, VnEntropyMP, MutualInfoMP)) for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    gradient_analysis(tape, grad_fn=hadamard_grad)
+    method = "analytic"
+    diff_methods = grad_method_validation(method, tape)
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    argnum = [i for i, dm in method_map.items() if dm == "A"]
+
+    if device_wires and len(tape.wires) == len(device_wires):
+        raise qml.QuantumFunctionError("The device has no free wire for the auxiliary wire.")
+
+    # Get default for aux_wire
+    if aux_wire is None:
+        aux_wire = _get_aux_wire(aux_wire, tape, device_wires)
+    elif aux_wire[0] in tape.wires:
+        raise qml.QuantumFunctionError("The auxiliary wire is already used.")
+    elif aux_wire[0] not in device_wires:
+        raise qml.QuantumFunctionError("The requested aux_wire does not exist on the used device.")
```

Comment:
For consistency with other errors:
```suggestion
        raise qml.QuantumFunctionError("The requested auxiliary wire does not exist on the used device.")
```